src/tasks/getsubscriptiondetails/taskgetsubscriptiondetails.cpp

```diff
@@ -0,0 +1,117 @@
+  MVPN_COUNT_CTOR(TaskGetSubscriptionDetails);
+}
+
+TaskGetSubscriptionDetails::~TaskGetSubscriptionDetails() {
+  MVPN_COUNT_DTOR(TaskGetSubscriptionDetails);
+}
+
+void TaskGetSubscriptionDetails::run() {
+  logger.debug() << "run";
+
+  NetworkRequest* request =
+      NetworkRequest::createForGetSubscriptionDetails(this);
+
+  connect(
+      request, &NetworkRequest::requestFailed, this,
+      [this](QNetworkReply::NetworkError error, const QByteArray&) {
+        logger.error() << "Get subscription details failed" << error;
+
+        // Network request failed after authentication for a second time
+        if (m_authenticationInAppSession) {
+          MozillaVPN::instance()->errorHandle(ErrorHandler::toErrorType(error));
+          m_authenticationInAppSession->terminate();
+          return;
+        }
+
+        // User needs to (re)authenticate
+        if (error == QNetworkReply::AuthenticationRequiredError) {
+          needsAuthentication();
+          return;
+        }
+
+        MozillaVPN::instance()->errorHandle(ErrorHandler::toErrorType(error));
+        emit completed();
+      });
+
+  connect(request, &NetworkRequest::requestCompleted, this,
+          [this](const QByteArray& data) {
+            logger.debug() << "Get subscription details completed" << data;
+
+            MozillaVPN::instance()->subscriptionDetailsFetched(data);
+
+            if (m_authenticationInAppSession) {
+              m_authenticationInAppSession->terminate();
+              return;
+            }
+            emit completed();
+          });
+}
+
+void TaskGetSubscriptionDetails::needsAuthentication() {
+  Q_ASSERT(!m_authenticationInAppSession);
+
+  QByteArray pkceCodeVerifier;
+  QByteArray pkceCodeChallenge;
+  AuthenticationListener::generatePkceCodes(pkceCodeVerifier,
+                                            pkceCodeChallenge);
+  Q_ASSERT(!pkceCodeVerifier.isEmpty() && !pkceCodeChallenge.isEmpty());
+
+  m_authenticationInAppSession = new AuthenticationInAppSession(this);
+
+  connect(m_authenticationInAppSession, &AuthenticationInAppSession::terminated,
+          this, &Task::completed);
+
+  connect(m_authenticationInAppSession, &AuthenticationInAppSession::completed,
+          this, [this, pkceCodeVerifier](const QString& pkceCodeSuccess) {
+            logger.debug() << "Authentication completed with code:"
+                           << pkceCodeSuccess;
+
+            NetworkRequest* request =
+                NetworkRequest::createForAuthenticationVerification(
+                    this, pkceCodeSuccess, pkceCodeVerifier);
+
+            connect(
+                request, &NetworkRequest::requestFailed, this,
+                [this](QNetworkReply::NetworkError error, const QByteArray&) {
+                  logger.error()
+                      << "Failed to complete the authentication" << error;
+                  m_authenticationInAppSession->terminate();
```

Comment:
What does it happen to this `Task` when this happens? We need to `complete` it somehow.

Reply:
What about if you connect a method to `NetworkRequest::destroyed` and you check the logic there?

Reply:
If the auth session it terminated completing the `Task` should be handled by `connect(m_authenticationInAppSession, &AuthenticationInAppSession::terminated, this, &Task::completed);` ([L81-82](https://github.com/mozilla-mobile/mozilla-vpn-client/pull/3538/files#diff-0de758d86f6de3f1fc61b369c831dfc6c1f3e4458dff91ab3a4312f604b2f8c6R81-R82)) if I’m not mistaken?